src/rgw/services/svc_mdlog.cc

```diff
@@ -313,7 +312,20 @@ Cursor RGWSI_MDLog::init_oldest_log_period()
   auto cursor = period_history->lookup(state.oldest_realm_epoch);
   if (cursor) {
     return cursor;
+  } else {
+    cursor = find_oldest_period();
+    state.oldest_realm_epoch = cursor.get_epoch();
+    state.oldest_period_id = cursor.get_period().get_id();
+    ldout(cct, 10) << "rewriting mdlog history" << dendl;
+    ret = write_history(state, &objv);
+    if (ret < 0 && ret != -EEXIST) {
```

Comment:
since this is writing with exclusive=false, it will overwrite the object instead of failing with EEXIST. so you don't need the special case for EEXIST here

however, since we're reusing the &objv from `read_history()`, we may get an ECANCELED error if another radosgw raced to write a newer version of this object. in that case, we either need to re-read the history and redo the lookup (basically rerun all of the logic in `init_oldest_log_period()`), or ignore the ECANCELED error and return the cursor as is. i _think_ the latter is okay, because a racing gateway should have seen the same period history

does that make sense?

Reply:
Yes, I have updated with the latter solution. Thanks.